### PR TITLE
Fix wording for expiring API keys email notification

### PR DESCRIPTION
**Pull request by shishirx34**
Fixes NuGet/NuGetGallery#3485
With the addition of Scoped API keys, we have a multiple API keys and the credential expiring job used to notify only about the single API key. This PR fixes this Job to send mail for expiring API keys(about to expire) and expired API keys. 

Sample message from this job looks like:
Expiring:
```
Subject: [NuGet Gallery] Your API key is about to expire

Hi user123,

We wanted to inform you that the following API key(s) on NuGet Gallery will expire soon:

• Full access API key – expires in 4 day(s).

• Push only key – expires in 3 day(s).

Visit https://nuget.localtest.me/account to generate a new API key(s) so that you can continue pushing packages using them.

Best regards,

NuGet Gallery
```

```
Subject: [NuGet Gallery] Your API key has expired

Hi user123,

We wanted to inform you that the following API key(s) on NuGet Gallery have expired.

• Api Key for Package1 – has expired.

Visit https://nuget.localtest.me/account to generate a new API key(s) so that you can continue pushing packages.

Best regards,

NuGet Gallery
```

### CODE REVIEW

File: src/Gallery.CredentialExpiration/Job.cs [hunk 1]
```diff
@@ -118,18 +119,47 @@ public override bool Init(IDictionary<string, string> jobArgsDictionary)
                         expiredCredentials.Count);
                 }
 
+                // Add default description for non-scoped API keys
+                expiredCredentials
```
Comment (scottbommarito) [inline src/Gallery.CredentialExpiration/Job.cs]: Does the gallery have code similar to this with regards to displaying the `Full access API key` description? Should we perform a migration that updates all empty description values to `Full access API key`?

File: src/Gallery.CredentialExpiration/Job.cs [hunk 2]
```diff
@@ -118,18 +119,47 @@ public override bool Init(IDictionary<string, string> jobArgsDictionary)
...
+                        await HandleExpiredCredentialEmail(username, expiredCredentialList, jobRunTime, expired: true);
+
+                        _logger.LogInformation("Handled expired/expiring credential for user {Username}.", username);
+                        _contactedUsers.AddOrUpdate(username, jobRunTime, (s, offset) => jobRunTime);
                     }
                     else
                     {
-                        _logger.LogDebug("Skipping expired credential for user {Username} - already handled today.",
-                            expiredCredential.Username);
+                        _logger.LogDebug("Skipping expired credential for user {Username} - already handled in last {days}.",
```
Comment (scottbommarito) [inline src/Gallery.CredentialExpiration/Job.cs]: Might be useful to print precisely when the expired/expiring credential was handled and not just the range of time.

File: src/Gallery.CredentialExpiration/Job.cs [hunk 3]
```diff
@@ -176,23 +219,30 @@ public override bool Init(IDictionary<string, string> jobArgsDictionary)
...
                 }
-
-                _logger.LogInformation("Handled expired credential for user {Username}.", expiredCredential.Username);
-
-                _contactedUsers.AddOrUpdate(expiredCredential.Username, jobRunTime, (s, offset) => jobRunTime);
             }
             catch (SmtpFailedRecipientException ex)
             {
-                _logger.LogWarning("Failed to handle expired credential for user {Username} - recipient failed.", expiredCredential.Username, ex);
+                _logger.LogWarning("Failed to handle expired credential for user {Username} - recipient failed! {Exception}", username, ex);
```
Comment (scottbommarito) [inline src/Gallery.CredentialExpiration/Job.cs]: If you want this to be consistent with the `Handling {Expired} credential(s) for user {Username} (Keys: {Descriptions})...` message above, you should make this message `Failed to handle {Expired} credential for user {Username} - recipient failed!`

File: src/Gallery.CredentialExpiration/Job.cs [hunk 4]
```diff
@@ -118,18 +119,47 @@ public override bool Init(IDictionary<string, string> jobArgsDictionary)
                         expiredCredentials.Count);
                 }
 
+                // Add default description for non-scoped API keys
+                expiredCredentials
+                    .Where(cred => (string.IsNullOrEmpty(cred.Description) || cred.IsNonScopedApiKey()))
```
Comment (scottbommarito) [inline src/Gallery.CredentialExpiration/Job.cs]: We may be masking actual descriptions if we keep the `cred.IsNonScopedApiKey()` condition. I would make it just `string.IsNullOrEmpty(cred.Description)`.

File: src/Gallery.CredentialExpiration/Strings.resx [hunk 5]
```diff
@@ -118,36 +118,40 @@
     <value>System.Resources.ResXResourceWriter, System.Windows.Forms, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089</value>
   </resheader>
   <data name="ExpiredEmailBody" xml:space="preserve">
-    <value>Hi,
+    <value>Hi {0},
 
-We wanted to inform you that your API key on {0} has expired.
+We wanted to inform you that the following API key(s) on {1} have expired.
```
Comment (scottbommarito) [inline src/Gallery.CredentialExpiration/Strings.resx]: Could we make this either `API key on {1} has expired` or `API keys on {1} have expired` based on the number of keys?

File: src/Gallery.CredentialExpiration/Strings.resx [hunk 6]
```diff
@@ -118,36 +118,40 @@
...
 
-Visit {1} to generate a new API key so that you can continue pushing packages.
+Visit {3} to generate a new API key(s) so that you can continue pushing packages using them.
 
 Best regards,
 
-{0}</value>
+{1}</value>
   </data>
   <data name="GetExpiredCredentialsQuery" xml:space="preserve">
```
Comment (scottbommarito) [inline src/Gallery.CredentialExpiration/Strings.resx]: Seems strange to have a SQL query in a `Strings.resx`. Most of our jobs with SQL queries have them as string constants in code or stored procedures. Can we move this query into one of those?

File: src/Gallery.CredentialExpiration/Job.cs [hunk 7]
```diff
@@ -118,18 +119,47 @@ public override bool Init(IDictionary<string, string> jobArgsDictionary)
...
+                        .ToList();
+                    var expiredCredentialList = credentialList
+                        .Where(x => (x.Expires - jobRunTime).TotalDays <= 0)
+                        .ToList();
+
+                    if (!_contactedUsers.ContainsKey(username))
                     {
-                        await HandleExpiredCredentialEmail(expiredCredential, jobRunTime);
+                        // send expiring API keys email notification
+                        await HandleExpiredCredentialEmail(username, expiringCredendtialList, jobRunTime, expired: false);
```
Comment (shishirx34) [inline src/Gallery.CredentialExpiration/Job.cs]: This is by design. The wordings are different for subject and email and it is not worth it merging these two. @anangaur agrees.

**Comment by skofman1** [comment]
>                 _logger.LogCritical("Failed to initialize job! {Exception}", exception);

Can you verify this logging is still valid? Look at this PR by Xavier: [https://github.com/NuGet/NuGet.Jobs/pull/103/files](https://github.com/NuGet/NuGet.Jobs/pull/103/files)

---
Refers to: src/Gallery.CredentialExpiration/Job.cs:76 in c5e9e45. [](commit_id = c5e9e45be52125a7c10101cdb99524595e33c2f1, deletion_comment = False)

File: src/Gallery.CredentialExpiration/Job.cs (hunk 4 above)
Comment (shishirx34) [inline src/Gallery.CredentialExpiration/Job.cs]: Currently there aren't descriptions on fathered in API keys. Gallery adds it at view level, but I guess no need for checking the scoping. Will change this.

**Comment by skofman1** [comment]
>                 if (File.Exists(_cursorFile))

will deployment delete this file... we might spam users. A more robust solution will be to have it in blob storage

---
Refers to: src/Gallery.CredentialExpiration/Job.cs:91 in c5e9e45. [](commit_id = c5e9e45be52125a7c10101cdb99524595e33c2f1, deletion_comment = False)

File: src/Gallery.CredentialExpiration/Job.cs (hunk 1 above)
Comment (shishirx34) [inline src/Gallery.CredentialExpiration/Job.cs]: Yes it does. May be we should.

File: src/Gallery.CredentialExpiration/Job.cs [hunk 8]
```diff
@@ -118,18 +119,47 @@ public override bool Init(IDictionary<string, string> jobArgsDictionary)
...
+
+                        // send expired API keys email notification
+                        await HandleExpiredCredentialEmail(username, expiredCredentialList, jobRunTime, expired: true);
+
+                        _logger.LogInformation("Handled expired/expiring credential for user {Username}.", username);
```
Comment (shishirx34) [inline src/Gallery.CredentialExpiration/Job.cs]: I don't think this is necessary. I particularly don't like logging this either, but keeping it consistent the way it was. Main thing should be logging in case it fails, which it does. #wontfix

File: src/Gallery.CredentialExpiration/Job.cs (hunk 2 above)
Comment (shishirx34) [inline src/Gallery.CredentialExpiration/Job.cs]: Not worth it. It was handled in last few days is a fine reason. It doesn't give us any useful information to take any action. This is way too much logging already.

File: src/Gallery.CredentialExpiration/Strings.resx (hunk 5 above)
Comment (shishirx34) [inline src/Gallery.CredentialExpiration/Strings.resx]: I don't think it is useful, this is over engineering it, when its perfectly clear what a suffix of (s) means.

File: src/Gallery.CredentialExpiration/Models/ExpiredCredentialData.cs [hunk 9]
```diff
@@ -5,12 +5,25 @@
 
 namespace Gallery.CredentialExpiration.Models
 {
+    public static class Constants
+    {
+        public const string ApiKeyV1 = "apikey.v1";
+        public const string ApiKeyV2 = "apikey.v2";
```
Comment (skofman1) [inline src/Gallery.CredentialExpiration/Models/ExpiredCredentialData.cs]: this already exists in NuGetGallery.Core package. Consider referencing it: [https://dotnet.myget.org/feed/nuget-build/package/nuget/NuGetGallery.Core](https://dotnet.myget.org/feed/nuget-build/package/nuget/NuGetGallery.Core)

File: src/Gallery.CredentialExpiration/Job.cs (hunk 2 above)
Comment (scottbommarito) [inline src/Gallery.CredentialExpiration/Job.cs]: I think it could be pretty useful with debugging. If we logged the timestamp that the credential was handled we could refer back to it in our logs.

**Comment by skofman1** [comment]
>                 _logger.LogCritical("Job run failed! {Exception}", ex);

same, please verify 

---
Refers to: src/Gallery.CredentialExpiration/Job.cs:168 in c5e9e45. [](commit_id = c5e9e45be52125a7c10101cdb99524595e33c2f1, deletion_comment = False)

File: src/Gallery.CredentialExpiration/Job.cs (hunk 2 above)
Comment (shishirx34) [inline src/Gallery.CredentialExpiration/Job.cs]: This is so much low pri its funny, I honestly don't think you would ever end up debugging why only a given user wasn't sent an expiry email. I ended up adding it nevertheless, #justforyou :D

File: src/Gallery.CredentialExpiration/Strings.resx (hunk 6 above)
Comment (shishirx34) [inline src/Gallery.CredentialExpiration/Strings.resx]: I agree, but I don't want to make unnecessary changes to this PR.

File: src/Gallery.CredentialExpiration/Models/ExpiredCredentialData.cs (hunk 9 above)
Comment (shishirx34) [inline src/Gallery.CredentialExpiration/Models/ExpiredCredentialData.cs]: I ended up removing these, not really needed as such. For now, the behavior is if the description is missing, use the default string below.

**Comment by shishirx34** [comment]
@skofman1 - About cursor file comment, I couldn't reply to the conversation about cursor file above for some reason. I think the deployment will delete this file. And its possible that we spam a few users once. I think moving to a blob storage, while good solution, is a bigger undertaking than necessary for this PR, we rarely do deployment of this job. If you still think it would be worth it we can create a separate task for it.

**Comment by skofman1** [comment]
@shishirx34 , lets understand how many users we are going to spam, before we go ahead with this. If there are too many such users, we should consider manual intervention in the deployment, to backup the cursor file, and later to restore it.
Regarding task for next sprint, agree - lets create an issue.

**Comment by shishirx34** [comment]
Created task https://github.com/NuGet/Engineering/issues/256. I will make a note for deployment to check the cursor file for backup after merging this PR.

File: src/Gallery.CredentialExpiration/Job.cs [hunk 10]
```diff
@@ -118,18 +119,47 @@ public override bool Init(IDictionary<string, string> jobArgsDictionary)
...
+                        _contactedUsers.AddOrUpdate(username, jobRunTime, (s, offset) => jobRunTime);
```
Comment (cristinamanum) [inline src/Gallery.CredentialExpiration/Job.cs]: Observation 1: If the first HandleExpiredCredential pass and the second throws that _contactedUser will not be updated - this means that the users may receive more emails than necessary.
Observation 2: If send email fails with handled SmtpFailedRecipientException  exception for only one of the above calls, the user will miss a notification.

Both cases may be acceptable, but I think it would be nice to add few comments about the expected behavior.

Comment (shishirx34) [inline src/Gallery.CredentialExpiration/Job.cs]: Good point. Moving the add contact users info after sending email, second time it would be overwritten, which is fine I think. If first one fails, user will never get an email. If second call fails, the user will get first email, second will be skipped, and won't get multiple mails again.

**Comment by shishirx34** [comment]
@skofman1 - The logging is still valid. The changes to Xavier's PR only addresses logging for Stats.AzureCdnLogs, and can't be applied here without moving it out to common code.

**Comment by skofman1** [comment]
What I meant is that @xavierdecoster discovered that although LogCritical calls were building correctly, they didn't actually output the exception. So @xavierdecoster replaced:
_logger.LogCritical("Failed to initialize job! {Exception}", exception);

with: 
_logger.LogCritical(LogEvents.JobInitFailed, exception, "Failed to initialize job!");

So I'm wondering if exception logging works correctly in this job.

File: src/Gallery.CredentialExpiration/Job.cs [hunk 11]
```diff
@@ -177,22 +218,41 @@ public override bool Init(IDictionary<string, string> jobArgsDictionary)
...
+                    expired ? "expired" : "expiring",
+                    username);
 
-                _contactedUsers.AddOrUpdate(expiredCredential.Username, jobRunTime, (s, offset) => jobRunTime);
+                _contactedUsers.AddOrUpdate(username, jobRunTime, (s, offset) => jobRunTime);
             }
             catch (SmtpFailedRecipientException ex)
             {
-                _logger.LogWarning("Failed to handle expired credential for user {Username} - recipient failed.", expiredCredential.Username, ex);
+                _logger.LogWarning("Failed to handle {Expired} credential for user {Username} - recipient failed! {Exception}",
```
Comment (xavierdecoster) [inline src/Gallery.CredentialExpiration/Job.cs]: This should become:
```
_logger.LogWarning(<eventId>, <exception>, <msgTemplate>, <msgParams>);
```

See https://docs.microsoft.com/en-us/aspnet/core/fundamentals/logging#logging-exceptions

File: src/Gallery.CredentialExpiration/Job.cs [hunk 12]
```diff
@@ -177,22 +218,41 @@ public override bool Init(IDictionary<string, string> jobArgsDictionary)
...
+                    expired ? "expired" : "expiring",
+                    username,
+                    ex);
             }
             catch (Exception ex)
             {
-                _logger.LogCritical("Failed to handle expired credential for user {Username}.", expiredCredential.Username, ex);
+                _logger.LogCritical("Failed to handle {Expired} credential for user {Username}! {Exception}",
```
Comment (xavierdecoster) [inline src/Gallery.CredentialExpiration/Job.cs]: This should become:
```
_logger.LogCritical(<eventId>, <exception>, <msgTemplate>, <msgParams>);
```

See https://docs.microsoft.com/en-us/aspnet/core/fundamentals/logging#logging-exceptions

**Comment by shishirx34** [comment]
Fixed the exceptions logging for this job. I would have liked LogEvents to be moved to common place for Nuget.Jobs and use it through out all our jobs, however that is outside the scope for this PR, hence duplicated the log events behavior for Gallery.ExpireCredentials job. I have created another bug, which should holistically fix the issue. https://github.com/NuGet/NuGetGallery/issues/3574

**Comment by shishirx34** [comment]
@xavierdecoster - sign off?
